### PR TITLE
[`ESM`] Add support for sdpa.

### PR DESCRIPTION
**Pull request by coekfung**
# What does this PR do?

Add support for SDPA (scaled dot product attention) for ESM. More context in https://github.com/huggingface/transformers/pull/28802 (And this pr mainly reused the code from this pr as the ESM is Bert-based model) and https://github.com/huggingface/transformers/issues/28005 .

This is my first time contributing to this project, please point out if there is any mistakes.

And revert a change in https://github.com/huggingface/transformers/pull/29329 as the dtype-mismatching issue for bitsandbytes is actually caused by the rotary embedding.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker

### CODE REVIEW

**Comment by Rocketknight1** [comment]
Thank you for this! Rather than skipping the SDPA test, though, can you write even a simple test that uses the SDPA path? It's okay if it can't compare hidden states deeply because of issues in the ESM model, but if it could compare that output logits are similar that'd give us a lot more confidence in the SDPA code!

**Comment by wzf03** [comment]
> Thank you for this! Rather than skipping the SDPA test, though, can you write even a simple test that uses the SDPA path? It's okay if it can't compare hidden states deeply because of issues in the ESM model, but if it could compare that output logits are similar that'd give us a lot more confidence in the SDPA code!

Thanks for your reply, I will add relevant test cases soon.

**Comment by wzf03** [comment]
> Thank you for this! Rather than skipping the SDPA test, though, can you write even a simple test that uses the SDPA path? It's okay if it can't compare hidden states deeply because of issues in the ESM model, but if it could compare that output logits are similar that'd give us a lot more confidence in the SDPA code!

@Rocketknight1 Hello, the sdpa inference tests for ESMFold has been added. Could you please review it?

File: src/transformers/models/esm/modeling_esm.py [hunk 1]
```diff
@@ -102,12 +106,20 @@ def _update_cos_sin_tables(self, x, seq_dimension=2):
...
+                freqs = torch.outer(t.float(), self.inv_freq.float())
+                emb = torch.cat((freqs, freqs), dim=-1).to(x.device)
+
+                cos = emb.cos()[None, None, :, :]
+                sin = emb.sin()[None, None, :, :]
 
-            self._cos_cached = emb.cos()[None, None, :, :]
-            self._sin_cached = emb.sin()[None, None, :, :]
+            self._cos_cached = cos.to(dtype=x.dtype)
+            self._sin_cached = sin.to(dtype=x.dtype)
```
Comment (Rocketknight1) [inline src/transformers/models/esm/modeling_esm.py:122]: We actually had issues with this section when porting ESM originally - ESM was trained with `float16` `self.inv_freq` and we had to be very careful not to clobber that with recomputed `float32` values, or we got incorrect outputs. I don't think this code replaces the values in `self.inv_freq`, so it's okay, but just making a note to be careful there!

**Review by Rocketknight1 — APPROVED** [review]
Overall this looks like a good SDPA addition to me! I'll also set up slow tests in a sec.

**Comment by Rocketknight1** [comment]
Hi @wzf03, I ran the full test suite for ESM and I'm seeing one or two test failures. Can you see if you can reproduce those locally? They may just be flaky tests, but it might also be caused by changes in this PR.

**Comment by wzf03** [comment]
> Hi @wzf03, I ran the full test suite for ESM and I'm seeing one or two test failures. Can you see if you can reproduce those locally? They may just be flaky tests, but it might also be caused by changes in this PR.

I found the test failures were due to the device mismatching of the `input_ids` (on cpu) and the `model` (on cuda). It can be reproduced locally with the current latest master branch of `accelerate@29be4788629b772a3b722076e433b5b3b5c85da3`. But in may original test environment with `accelerate==1.1.1`, everything works well.

I will report this to `accelerate` later.

**Comment by wzf03** [comment]
Hello @Rocketknight1 , I made a quick fix according to other model's test, the test cases should work normally now.

**Comment by wzf03** [comment]
@ArthurZucker @LysandreJik Hello! Can you please help review this pr?

**Review by ArthurZucker — COMMENTED** [review]
Hey super sorry for the delay, waited a bit because #35235 changes the interface! Do you mind updating this PR ? Hope it's not too much of a burden! 😿

**Comment by wzf03** [comment]
> Hey super sorry for the delay, waited a bit because #35235 changes the interface! Do you mind updating this PR ? Hope it's not too much of a burden! 😿

Sure, I will do it soon.

**Comment by pstjohn** [comment]
Not sure if this is still active, but I have a similar PR in #38023 to add flash attention 2 to ESM

**Comment by wzf03** [comment]
@ArthurZucker @Rocketknight1 Sorry for the late update. I have merged the sdpa support into the new codebase, can you help review this?

File: src/transformers/models/esm/modeling_esm.py [hunk 2]
```diff
@@ -387,6 +405,113 @@ def forward(
...
+                key_layer = torch.cat([past_key_value[0], key_layer], dim=2)
+                value_layer = torch.cat([past_key_value[1], value_layer], dim=2)
+
+        # Scale the query for rotary embeddings
+        query_layer = query_layer * self.attention_head_size**-0.5
+
+        if self.is_decoder:
+            # if cross_attention save Tuple(torch.Tensor, torch.Tensor) of all cross attention key/value_states.
+            # Further calls to cross_attention layer can then reuse all cross-attention
+            # key/value_states (first "if" case)
```
Comment (ArthurZucker) [inline src/transformers/models/esm/modeling_esm.py:470]: I think you are still missing the spot : we don't need 3 different classes anymore https://github.com/huggingface/transformers/blob/tp-cb/src/transformers/models/llama/modeling_llama.py#L249-L249

**Review by ArthurZucker — COMMENTED** [review]
thanks for updating!

**Comment by pstjohn** [comment]
looks like #38751 broke the existing flash_attention_2 implementation for ESM-2 as well, so we're back to only eager being supported.

**Comment by ArthurZucker** [comment]
cc @zucchini-nlp !
